File: web/src/lib/capacity.test.ts
import { describe, expect, it } from "vitest";
import { calculateCapacity } from "./capacity";
import type { Node, Resources } from "@/types";

const resources: Resources = { cpu_millis: 1000, memory_bytes: 2_000, gpu: { count: 1, min_vram_bytes: 6_000 } };
const node = (overrides: Partial<Node>): Node => ({ id: "node-a", name: "Node A", status: "ONLINE", agent_version: "test", architecture: "amd64", docker_version: "test", cpu_total_millis: 8000, cpu_allocated_millis: 2000, memory_total_bytes: 16000, memory_allocated_bytes: 4000, workspace_free_bytes: 10000, labels: { zone: "lab" }, gpus: [{ uuid: "gpu-a", model: "RTX", vram_bytes: 8000, allocated: false }], gpu_discovery: { status: "available" }, last_heartbeat: new Date().toISOString(), ...overrides });

describe("calculateCapacity", () => {
  it("uses capacity from one compatible node", () => { const result = calculateCapacity([node({})], { zone: "lab" }, resources); expect(result.feasible).toHaveLength(1); expect(result.cpuMax).toBe(6000); expect(result.memoryMax).toBe(12000); expect(result.gpuMax).toBe(1); expect(result.vramMax).toBe(8000); });
  it("never adds GPUs across nodes", () => { const result = calculateCapacity([node({ id: "a" }), node({ id: "b" })], {}, { ...resources, gpu: { count: 2, min_vram_bytes: 0 } }); expect(result.feasible).toHaveLength(0); expect(result.gpuMax).toBe(1); });
  it("filters selectors before calculating limits", () => { const result = calculateCapacity([node({})], { zone: "remote" }, resources); expect(result.nodes).toHaveLength(0); expect(result.cpuMax).toBe(0); });
});

File: internal/capacity/capacity_test.go
package capacity

import (
	"testing"

	"github.com/jobdock/jobdock/internal/domain"
)

func TestAccountReservations(t *testing.T) {
	nodes := []domain.Node{{ID: "node-1", GPUs: []domain.GPU{{UUID: "GPU-1"}, {UUID: "GPU-2"}}}}
	jobs := []domain.Job{{AssignedNodeID: "node-1", Status: domain.JobRunning, Spec: domain.JobSpec{Resources: domain.Resources{CPUMillis: 1500, MemoryBytes: 2 << 30}}}}
	Account(nodes, jobs, map[string]map[string]bool{"node-1": {"GPU-2": true}})
	if nodes[0].CPUAllocatedMillis != 1500 || nodes[0].MemoryAllocatedBytes != 2<<30 {
		t.Fatalf("unexpected allocation: %#v", nodes[0])
	}
	if nodes[0].GPUs[0].Allocated || !nodes[0].GPUs[1].Allocated {
		t.Fatalf("unexpected GPU allocation: %#v", nodes[0].GPUs)
	}
}
